File: iosApp/iosApp/features/product_detail/bdui/BduiRenderer.swift
//
//  BduiRenderer.swift
//  iosApp
//

import SwiftUI
import ComposeApp

/// Action context is an OPAQUE Kotlin `ActionContext` (not a Swift `Dictionary`) on purpose: it must
/// never be bridged to/from `[String: JsonElement]`, or the round trip yields a dangling Kotlin `Map`
/// that crashes on `Map#get`. The wrapped `JsonObject` is only ever unwrapped in Kotlin
/// (`ProductDetailViewStore.dispatchBduiAction`). See `ActionContext` in shared.
typealias BduiActionHandler = (String, ActionContext) -> Void

/// Resolves contextBindings from BindData and merges extra literal values into an opaque ActionContext.
/// Literal keys (targetSlotId, url, replace — shared `LITERAL_CONTEXT_KEYS`) are authored verbatim,
/// not as bind paths, so they are pulled out of the path-resolved bindings and passed through as-is.
func buildActionContext(bindings: [String: String], data: BindData, extra: [String: String] = [:]) -> ActionContext {
    let literalKeys = BduiConstants.shared.LITERAL_CONTEXT_KEYS
    var pathBindings = bindings
    var literals = extra
    for key in literalKeys {
        if let value = pathBindings.removeValue(forKey: key) {
            literals[key] = value
        }
    }
    return data.resolveActionContext(bindings: pathBindings, extra: literals)
}

/// Replaced slots in scope for the current subtree, keyed by `targetSlotId`. Empty by default and
/// re-scoped to empty inside a replacement so the standalone layout never re-triggers a swap.
private struct BduiReplacedSlotsKey: EnvironmentKey {
    static let defaultValue: [String: ResolvedReplace] = [:]
}

extension EnvironmentValues {
    var bduiReplacedSlots: [String: ResolvedReplace] {
        get { self[BduiReplacedSlotsKey.self] }
        set { self[BduiReplacedSlotsKey.self] = newValue }
    }
}

struct BduiNodeView: View {
    let node: UiNode
    let data: BindData
    let onAction: BduiActionHandler
    @Environment(\.bduiReplacedSlots) private var replacedSlots

    var body: some View {
        if let replacement = BduiReplaceResolver.shared.replacementFor(node: node, replacedSlots: replacedSlots) {
            // Standalone-scope replace: render the replacement layout against its own `values`
            // instead of the original subtree + product bindData. Clear replacedSlots for the
            // subtree so the standalone layout can't recursively re-trigger a swap.
            //
            // AnyView is REQUIRED here, not cosmetic: this branch embeds `BduiNodeView` directly
            // (not via a ForEach), so without erasure SwiftUI's static `_viewListCount` would
            // recurse `_ConditionalContent` → BduiNodeView → _ConditionalContent → … forever and
            // hang the render thread (the body never commits, so the loader stays on screen).
            // AnyView gives this branch a dynamic count and terminates the static recursion. The
            // normal subtree (else branch) is safe because it only recurses through ForEach.
            AnyView(
                BduiNodeView(
                    node: replacement.node,
                    data: replacement.bindData,
                    onAction: onAction
                )
                .environment(\.bduiReplacedSlots, [:])
            )
        } else {
            renderedBody
                .modifier(NodeTapModifier(node: node, data: data, onAction: onAction))
                .modifier(NodeLayoutModifier(node: node))
        }
    }

    @ViewBuilder
    private var renderedBody: some View {
        switch onEnum(of: node) {
        case .column(let n): ColumnView(node: n, data: data, onAction: onAction)
        case .row(let n): RowView(node: n, data: data, onAction: onAction)
        case .section(let n): SectionView(node: n, data: data, onAction: onAction)
        case .text(let n): TextNodeView(node: n, data: data)
        case .image(let n): ImageNodeView(node: n, data: data)
        case .imageGallery(let n): ImageGalleryNodeView(node: n, data: data)
        case .priceBlock(let n): PriceBlockView(node: n, data: data)
        case .specTable(let n): SpecTableView(node: n, data: data)
        case .sizeSelector(let n): SizeSelectorView(node: n, data: data, onAction: onAction)
        case .colorSwatchPicker(let n): ColorSwatchPickerView(node: n, data: data, onAction: onAction)
        case .button(let n): ButtonNodeView(node: n, data: data, onAction: onAction)
        case .spacer(let n): SpacerNodeView(node: n)
        case .divider: DividerNodeView()
        }
    }
}

/// Dispatches a node's optional `onTap` action. Applied inside NodeLayoutModifier so the tap
/// area matches the node's visual bounds. Children with their own tap handling (e.g. Button)
/// consume taps before this gesture, giving the innermost-handler-wins rule for free.
private struct NodeTapModifier: ViewModifier {
    let node: UiNode
    let data: BindData
    let onAction: BduiActionHandler

    func body(content: Content) -> some View {
        if let onTap = node.onTap {
            content
                .contentShape(Rectangle())
                .onTapGesture {
                    onAction(onTap.actionId, buildActionContext(bindings: onTap.contextBindings, data: data))
                }
        } else {
            content
        }
    }
}

/// Single-child layout that allocates `fraction * parentWidth` to its child and positions it
/// according to `alignment` within the full parent width. Unlike GeometryReader it correctly
/// reports the child's actual height to the parent, so ScrollView contentSize is never underestimated.
private struct FractionalWidthLayout: Layout {
    let fraction: Float
    let alignment: NodeAlignment?

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        guard let totalWidth = proposal.width, totalWidth > 0 else {
            return subviews.first?.sizeThatFits(.unspecified) ?? .zero
        }
        let childWidth = totalWidth * CGFloat(fraction)
        let childHeight = subviews.first?.sizeThatFits(
            ProposedViewSize(width: childWidth, height: proposal.height)
        ).height ?? 0
        return CGSize(width: totalWidth, height: childHeight)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        guard let subview = subviews.first else { return }
        let childWidth = bounds.width * CGFloat(fraction)
        let x: CGFloat
        switch alignment {
        case .center?: x = bounds.minX + (bounds.width - childWidth) / 2
        case .end?:    x = bounds.maxX - childWidth
        default:       x = bounds.minX
        }
        subview.place(
            at: CGPoint(x: x, y: bounds.minY),
            proposal: ProposedViewSize(width: childWidth, height: proposal.height)
        )
    }
}

private struct NodeLayoutModifier: ViewModifier {
    let node: UiNode

    func body(content: Content) -> some View {
        let widthFraction = node.widthFraction?.floatValue
        let alignment = node.alignment

        Group {
            if let w = widthFraction {
                FractionalWidthLayout(fraction: w, alignment: alignment) {
                    content
                }
            } else {
                aligned(content: content, alignment: alignment)
            }
        }
    }

    @ViewBuilder
    private func aligned(content: Content, alignment: NodeAlignment?) -> some View {
        switch alignment {
        case .start?:
            HStack { content; Spacer(minLength: 0) }
        case .center?:
            HStack { Spacer(minLength: 0); content; Spacer(minLength: 0) }
        case .end?:
            HStack { Spacer(minLength: 0); content }
        default:
            content
        }
    }
}

// MARK: - Token helpers

func spacing(_ s: Spacing) -> CGFloat {
    switch s {
    case .sm: return 4
    case .md: return 8
    case .lg: return 16
    case .xl: return 24
    default: return 8
    }
}

func spacerHeight(_ s: SpacerSize) -> CGFloat {
    switch s {
    case .sm: return 8
    case .md: return 16
    case .lg: return 24
    case .xl: return 40
    default: return 16
    }
}

func bduiFont(for style: TextStyle) -> Font {
    switch style {
    case .title: return .system(size: 24, weight: .semibold)
    case .subtitle: return .system(size: 18, weight: .medium)
    case .body: return .system(size: 14)
    case .caption: return .system(size: 12)
    default: return .body
    }
}

func textAlignment(for alignment: NodeAlignment?) -> TextAlignment {
    switch alignment {
    case .center?: return .center
    case .end?: return .trailing
    default: return .leading
    }
}

File: composeApp/src/androidMain/kotlin/org/example/fakeshop_clients/features/product_detail/presentation/bdui/BduiRenderer.kt
package org.example.fakeshop_clients.features.product_detail.presentation.bdui

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import org.example.fakeshop_clients.features.bdui.BduiConstants
import org.example.fakeshop_clients.features.bdui.domain.BduiReplaceResolver
import org.example.fakeshop_clients.features.bdui.domain.models.ResolvedReplace
import org.example.fakeshop_clients.features.bdui.domain.models.UiNode
import org.example.fakeshop_clients.features.bdui.domain.models.resolve
import org.example.fakeshop_clients.features.bdui.domain.models.tokens.NodeAlignment
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderButton
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderColorSwatchPicker
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderColumn
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderDivider
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderImage
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderImageGallery
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderPriceBlock
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderRow
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderSection
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderSizeSelector
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderSpacer
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderSpecTable
import org.example.fakeshop_clients.features.product_detail.presentation.bdui.nodes.RenderText

val LocalBduiActionHandler = compositionLocalOf<(String, JsonObject) -> Unit> { { _, _ -> } }

/**
 * Replaced slots in scope for the current subtree, keyed by `targetSlotId`. Empty by default
 * and re-scoped to empty inside a replacement so the standalone layout never re-triggers a swap.
 */
val LocalReplacedSlots = compositionLocalOf<Map<String, ResolvedReplace>> { emptyMap() }

fun buildActionContext(bindings: Map<String, String>, data: JsonObject): JsonObject =
    buildJsonObject {
        bindings.forEach { (key, value) ->
            if (key == BduiConstants.TARGET_SLOT_ID_KEY) {
                // `targetSlotId` is authored as a verbatim literal (replace action), not a
                // bind path — pass it through unresolved so the slot swap can find its target.
                put(key, value)
            } else {
                data.resolve(value)?.let { put(key, it) }
            }
        }
    }

@Composable
fun BduiRenderer(
    root: UiNode,
    data: JsonObject,
    replacedSlots: Map<String, ResolvedReplace> = emptyMap(),
    modifier: Modifier = Modifier
) {
    CompositionLocalProvider(LocalReplacedSlots provides replacedSlots) {
        RenderNode(root, data, modifier)
    }
}

@Composable
fun RenderNode(node: UiNode, data: JsonObject, modifier: Modifier = Modifier) {
    // Standalone-scope replace: if this container's slot was replaced, render the replacement
    // layout against its own `values` instead of the original subtree + product bindData.
    // Re-scope replacedSlots to empty so the replacement is self-contained (no recursive swap).
    val replacement = BduiReplaceResolver.replacementFor(node, LocalReplacedSlots.current)
    if (replacement != null) {
        CompositionLocalProvider(LocalReplacedSlots provides emptyMap()) {
            RenderNode(replacement.node, replacement.values, modifier)
        }
        return
    }

    val widthFraction = node.widthFraction
    val alignment = node.alignment

    if (alignment != null) {
        val contentAlignment = when (alignment) {
            NodeAlignment.start -> Alignment.CenterStart
            NodeAlignment.center -> Alignment.Center
            NodeAlignment.end -> Alignment.CenterEnd
        }
        Box(
            modifier = modifier.fillMaxWidth(),
            contentAlignment = contentAlignment
        ) {
            val innerWidthMod = if (widthFraction != null) Modifier.fillMaxWidth(widthFraction) else Modifier
            Box(modifier = innerWidthMod) {
                RenderNodeContent(node, data)
            }
        }
    } else {
        val widthMod = if (widthFraction != null) Modifier.fillMaxWidth(widthFraction) else Modifier
        Box(modifier = modifier.then(widthMod)) {
            RenderNodeContent(node, data)
        }
    }
}

@Composable
private fun RenderNodeContent(node: UiNode, data: JsonObject) {
    when (node) {
        is UiNode.Column -> RenderColumn(node, data)
        is UiNode.Row -> RenderRow(node, data)
        is UiNode.Section -> RenderSection(node, data)
        is UiNode.Text -> RenderText(node, data)
        is UiNode.Image -> RenderImage(node, data)
        is UiNode.ImageGallery -> RenderImageGallery(node, data)
        is UiNode.PriceBlock -> RenderPriceBlock(node, data)
        is UiNode.SpecTable -> RenderSpecTable(node, data)
        is UiNode.SizeSelector -> RenderSizeSelector(node, data)
        is UiNode.ColorSwatchPicker -> RenderColorSwatchPicker(node, data)
        is UiNode.Button -> RenderButton(node, data)
        is UiNode.Spacer -> RenderSpacer(node)
        is UiNode.Divider -> RenderDivider()
    }
}
